Monolithic kernel fixes

mcdermj opened the pull request:
This is a proposed fix to the issues described in #3011.  

The first patch is a cleanup patch because the probe function does not properly free the IRQ upon failure.  This can be relatively easily remedied by having it use the devm_ function to properly do resource management.

The second is the meat of the fix.  The real way to see if the clock is there is to restore the code to try to get it.  We just do this and then get the clock's name in the clock registration function.

This has been tested on both monolithic and module-based kernels to make sure it works correctly for both.  It seems to work fine with the hardware I have to test with here.  Commentary is always welcome.

If this looks acceptable, I'll get this committed upstream and resubmit a PR with the upstream commit ID.

Thanks for your patience and many apologies for breaking things for folks.

lategoodbye reviewed (commented):
In general devm_ functions are preferred, but there are corner cases regarding race conditions (IRQ is shared between 3 interfaces including VPU). So we must be careful to not replace a potential resource leak with a hard to debug race condition during driver unload. An alternative solution would be to move all the clk handling before the irq stuff. Please consider this only as a gut feeling.

lategoodbye reviewed (commented):
The clock fix looks good to me.

pelwell commented:
I'm happy with these patches, but I defer to @lategoodbye on the mysterious potential devm irq race condition.

pelwell commented:
I'm happy with these patches, but I defer to @lategoodbye on the mysterious potential devm irq race condition.

pelwell commented:
I'm happy with these patches, but I defer to @lategoodbye on the mysterious potential devm irq race condition.

mcdermj commented:
I’ll defer to @lategoodbye as well as that’s out of my area of expertise. I’ll respin the second patch with the IRQ registration code at the end and get it submitted for upstream review this evening.

mcdermj commented:
I submitted a fix upstream this evening.  As I stated before, I'll submit a PR when I have it accepted there.

mcdermj committed:
i2c: bcm2835: Move IRQ request after clock code in probe

Commit 4a5cfa39465cad25dd736d7ceba8a5d32eea4ecc upstream.

If any of the clock code in the probe fails and returns, the IRQ
will not be freed.  Moving the IRQ request to last allows it to
be freed on any errors further up in the probe function.  devm_
calls can apparently not be used because there are some potential
race conditions that will arise.

Fixes: bebff81fb8b9 ("i2c: bcm2835: Model Divider in CCF")
Signed-off-by: Annaliese McDermond <nh6z@nh6z.net>
Acked-by: Stefan Wahren <wahrenst@gmx.net>
Signed-off-by: Wolfram Sang <wsa@the-dreams.de>

mcdermj committed:
i2c: bcm2835: Ensure clock exists when probing

Commit 9de93b04df16b055824e3f1f13fedb90fbcf2e4f upstream.

Probe function fails to recognize that upstream clock actually
doesn't yet exist because clock driver has not been initialized.
Actually try to go get the clock and test for its existence
before trying to set up a downstream clock based upon it.

This fixes a bug that causes the i2c driver not to work with
monolithic kernels.

Fixes: bebff81fb8b9 ("i2c: bcm2835: Model Divider in CCF")
Signed-off-by: Annaliese McDermond <nh6z@nh6z.net>
Acked-by: Stefan Wahren <wahrenst@gmx.net>
Signed-off-by: Wolfram Sang <wsa@the-dreams.de>

mcdermj commented:
@pelwell This has been accepted upstream and seems to fix the issue.  Please feel free to let me know if there's something else I need to take care of.
